replica/database.cc

```diff
@@ -1790,6 +1790,8 @@ future<mutation> database::do_apply_counter_update(column_family& cf, const froz
                 transform_counter_updates_to_shards(m, mopt ? &*mopt : nullptr, cf.failed_counter_applies_to_memtable(), get_token_metadata().get_my_id());
                 tracing::trace(trace_state, "Applying counter update");
                 return this->apply_with_commitlog(cf, m, timeout);
+            }).then([] {
+                return utils::get_local_injector().inject("apply_counter_update_delay_5s", std::chrono::seconds(5));
             }).then([&m] {
                 return std::move(m);
             });
```

Comment:
Calling a `.then` on an unresolved future will allocate a continuation, so your change will most likely add an allocation for every processed counter update. I suggest reformulating it like this:

```c++
.then([&m] {
    if (utils::get_local_injector().enter("apply_counter_update_delay_5s")) {
        return seastar::sleep(std::chrono::seconds(5)).then([&m] {
            return std::move(m);
        });
    }
    return make_ready_future<mutation>(std::move(m));
});
```

If you write it this way, it should have little to no impact if the injection is not enabled, and the injection logic should be completely optimized out in release mode.

Reply:
> Calling a `.then` on an unresolved future will allocate a continuation, so your change will most likely add an allocation for every processed counter update.

But this continuation was already allocated before for the existing .then

In the common case, wouldn't the injector just return a ready future and the .then continuation to return the mutation will run sequentially without yielding?

Reply:
all in all this function deserves to be coroutinized and simplified this way.

Reply:
> > Calling a `.then` on an unresolved future will allocate a continuation, so your change will most likely add an allocation for every processed counter update.
> 
> But this continuation was already allocated before for the existing .then

Yes, but now there is one more `.then` we allocate one more continuation.

> 
> In the common case, wouldn't the injector just return a ready future and the .then continuation to return the mutation will run sequentially without yielding?

It doesn't work like that. If you have code like:

```c++
return do_something_that_yields().then([] {
    return 42;
});
```

...then it will allocate a continuation because `do_something_that_yields()` returns a future that is not resolved, and `.then` will incur an allocation.

If you change it to:

```c++
return do_something_that_yields().then([] {
    return utils::get_local_injector().inject("apply_counter_update_delay_5s", std::chrono::seconds(5));
}).then([] {
    return 42;
});
```

Then you will always have __two__ continuations, because you call `.then` twice on an unresolved future, even if the injector call evaluates to a no-op.

Reply:
The error injection and test don't have to be backported. So the series can be done as:
1. minimal and backportable fix
2. coroutinization
3. error injection test

Reply:
> The error injection and test don't have to be backported. So the series can be done as:
> 
> 1. minimal and backportable fix
> 2. coroutinization
> 3. error injection test

Makes sense to me, thanks. @mlitvk I suggest we go this route: let's only leave the (1) part in this PR and split out (2, 3) to another PR. We'll only mark this PR for backport but will skip backporting the second one.

Reply:
ok, I removed the error injection and test from this PR